core/src/main/scala/org/apache/spark/rdd/SkewedJoinRDD.scala

```diff
@@ -0,0 +1,345 @@
+
+    override def flatten(i: Iterator[(K, (Iterable[Chunk[L]], Iterable[Chunk[R]]))]) =
+      i flatMap {
+        case (key, pair) => {
+          if (pair._1.size == 0) {
+            for (chunk <- pair._2.iterator;
+                 v <- chunk
+            ) yield (key, (None, Some(v))): (K, (Option[L], Option[R]))
+          }
+          else if (pair._2.size == 0) {
+            for (chunk <- pair._1.iterator;
+                 v <- chunk
+            ) yield (key, (Some(v), None)): (K, (Option[L], Option[R]))
+          }
+          else if (pair._1.size < pair._2.size) {
+            yieldPair(pair._1, pair._2, key, (p1: L, p2: R) => (Some(p1), Some(p2)))
+          } else {
+            yieldPair(pair._2, pair._1, key, (p2: R, p1: L) => (Some(p1), Some(p2)))
+          }
+        }
+      }
+  }
+
+  private def yieldPair[K, OUT, IN, PAIR <: Product2[_, _]](
+      outer: Iterable[Chunk[OUT]], inner: Iterable[Chunk[IN]], key: K, toPair: (OUT, IN) => PAIR) =
+    for (
+      outerChunk <- outer.iterator;
+      innerChunk <- inner.iterator;
+      outerValue <- outerChunk;
+      innerValue <- innerChunk
+    ) yield (key, toPair(outerValue, innerValue))
+}
+
+private[spark] sealed trait JoinSplitDep extends Serializable
+
+private[spark] case class NarrowJoinSplitDep(
+    rdd: RDD[_],
+    splitIndex: Int,
+    var split: Partition) extends JoinSplitDep {
+
+  @throws(classOf[IOException])
+  private def writeObject(oos: ObjectOutputStream): Unit = Utils.tryOrIOException {
+    // Update the reference to parent split at the time of task serialization
+    split = rdd.partitions(splitIndex)
+    oos.defaultWriteObject()
+  }
+}
+
+private[spark] case class ShuffleJoinSplitDep(handle: ShuffleHandle) extends JoinSplitDep
+
+private[spark] class JoinPartition(idx: Int, val left: JoinSplitDep, val right: JoinSplitDep)
+  extends Partition with Serializable {
+  override val index: Int = idx
+
+  override def hashCode(): Int = idx
+}
+
+private[spark] class BufferWrapper private(val isChunkBuffer: Boolean, buffer: Iterable[Any])
+  extends Serializable {
+
+  def this(buffer: CompactBuffer[_]) {
+    this(false, buffer)
+  }
+
+  def this(buffer: ChunkBuffer[_]) {
+    this(true, buffer)
+  }
+
+  def getChunkBuffer[T: ClassTag]: ChunkBuffer[T] = buffer.asInstanceOf[ChunkBuffer[T]]
+
+  def getCompactBuffer[T: ClassTag]: CompactBuffer[T] = buffer.asInstanceOf[CompactBuffer[T]]
+
+  def asChunkIterable[T: ClassTag]: Iterable[Chunk[T]] = {
+    if (isChunkBuffer) {
+      getChunkBuffer[T]
+    }
+    else {
+      val buffer = getCompactBuffer[T]
+      if (buffer.isEmpty) {
+        Iterable[Chunk[T]]()
+      }
+      else {
+        Iterable(new Chunk[T](buffer))
+      }
+    }
+  }
+}
+
+private[spark] class SkewedJoinRDD[K, L, R, PAIR <: Product2[_, _]](
+    left: RDD[(K, L)], right: RDD[(K, R)], part: Partitioner, joinType: JoinType[K, L, R, PAIR])
+    (implicit kt: ClassTag[K], lt: ClassTag[L], rt: ClassTag[R], ord: Ordering[K])
```

Comment:
I think `skewedGroupby` also requires the key is ordering. Otherwise, using a hash comparator, because the hash conflicts, the may require loading all values of a key into memory in some special situation. 

E.g., assume k1 and k2 have the same hashcode,  if using a hash comparator, the data in two external files may be (assume value is 1):

file1: (k1, 1), (k2, 1), (otherKey, 1), ...
file2: (k2, 1),  (k2, 1), (k2, 1), ... ,.... (k1, 1), (otherKey, 1), ...

In file2, k1 is behind all `k2` (We use a hash comparator, so it's possible). Becuase we scan file1 at first, we will process `k1` before `k2`. But when processing k1 in `file2`, we need to load all `k2`'s values in `file2`, it may cause OOM.

If the key is ordering, we can avoid this extreme case.